### PR TITLE
impr: block scopes in the `nameRegistry`

### PR DESCRIPTION
**Pull request by cieplypolar**
This PR introduces finer-grained scoping to our `nameRegistry` by implementing block scopes.

It aims to:
- minimize variable shadowing
-  prepare for a transition from `esbuild` to `rolldown` in upcoming Vite 8

Additionally:
- updated the for ... of ... loop generation
- added edge cases for "for ... of ..."
- added future tests for rolldown

Below, I attach diagram of new `nameRegistry`
image

### CODE REVIEW

File: packages/typegpu/tests/tgsl/wgslGenerator.test.ts [hunk 1]
```diff
@@ -842,6 +874,112 @@ describe('wgslGenerator', () => {
...
+      }"
+    `);
+  });
+
+  it('handles "for ... of ..." internal index variable when "i" is the loop variable', () => {
+    const f = () => {
+      'use gpu';
+      const arr = [1, 2, 3];
+      let res = 0;
+      for (const i of arr) {
```
Comment (iwoplaza) [inline packages/typegpu/tests/tgsl/wgslGenerator.test.ts:935]: This was my mystery test case that, haha. Nice!

File: packages/typegpu/tests/examples/individual/jelly-switch.test.ts [hunk 2]
```diff
@@ -337,7 +337,7 @@ describe('jelly switch example', () => {
           return background;
         }
         var distanceFromOrigin = max(0f, intersection.tMin);
-        for (var i = 0; (i < 64i); i++) {
+        for (var i_1 = 0; (i_1 < 64i); i_1++) {
```
Comment (iwoplaza) [inline packages/typegpu/tests/examples/individual/jelly-switch.test.ts]: Why is this name changed?

File: packages/typegpu/tests/examples/individual/jelly-slider.test.ts [hunk 3]
```diff
@@ -533,7 +533,7 @@ describe('jelly-slider example', () => {
           return background;
         }
         var distanceFromOrigin = max(0f, intersection.tMin);
-        for (var i = 0; (i < 64i); i++) {
+        for (var i_1 = 0; (i_1 < 64i); i_1++) {
```
Comment (iwoplaza) [inline packages/typegpu/tests/examples/individual/jelly-slider.test.ts]: Same here

File: packages/typegpu/tests/examples/individual/jelly-switch.test.ts (hunk 2 above)
Comment (cieplypolar) [reply packages/typegpu/tests/examples/individual/jelly-switch.test.ts]: This change is related to how we generate regular for loops. Currently, we generate the loop header first and then create a new block for the body. Because of this, two consecutive loops cause a name collision
```ts
for (let i = 0; ...) {
...
}

for (let i = 7; ...) {
...
}
```
I’ll fix that first thing in the morning.

Comment (cieplypolar) [reply packages/typegpu/tests/examples/individual/jelly-switch.test.ts]: done

File: packages/typegpu/src/nameRegistry.ts [hunk 4]
```diff
@@ -458,31 +471,68 @@ abstract class NameRegistryImpl implements NameRegistry {
...
   }
 
   popFunctionScope(): void {
-    this.#usedFunctionScopeNamesStack.pop();
+    const functionScopeIndex = this.#scopeStack.findLastIndex((scope) =>
+      scope.type === 'functionScope'
+    );
+
+    if (functionScopeIndex === -1) {
+      return;
```
Comment (aleksanderkatan) [inline packages/typegpu/src/nameRegistry.ts]: nit
```suggestion
      throw new Error("Tried to pop function scope when no scope was present.");
```

File: packages/typegpu/src/tgsl/wgslGenerator.ts [hunk 5]
```diff
@@ -1142,18 +1142,23 @@ ${this.ctx.pre}else ${alternate}`;
...
-          condition ? this.typedExpression(condition, bool) : undefined,
-          update ? this.statement(update) : undefined,
-        ]);
-
-      const initStr = initStatement ? initStatement.slice(0, -1) : '';
-      const updateStr = updateStatement ? updateStatement.slice(0, -1) : '';
-
-      const bodyStr = this.block(blockifySingleStatement(body));
-      return stitch`${this.ctx.pre}for (${initStr}; ${conditionExpr}; ${updateStr}) ${bodyStr}`;
+      try {
```
Comment (aleksanderkatan) [inline packages/typegpu/src/tgsl/wgslGenerator.ts:1145]: What can fail here?

Comment (cieplypolar) [reply packages/typegpu/src/tgsl/wgslGenerator.ts:1145]:
```ts
init ? this.statement(init) : undefined,
condition ? this.typedExpression(condition, bool) : undefined,
update ? this.statement(update) : undefined,
```
Each one of them can throw.
